crates/bevy_ecs/src/world/archetype_invariants.rs

```diff
@@ -0,0 +1,707 @@
+        }
+    }
+
+    /// Constructs a new [`ArchetypeStatement::AllOf`] variant for all components stored in the bundle `B`.
+    #[inline]
+    pub const fn all_of() -> Self {
+        ArchetypeStatement::AllOf(PhantomData)
+    }
+
+    /// Constructs a new [`ArchetypeStatement::AnyOf`] variant for all components stored in the bundle `B`.
+    #[inline]
+    pub const fn any_of() -> Self {
+        ArchetypeStatement::AnyOf(PhantomData)
+    }
+
+    /// Constructs a new [`ArchetypeStatement::AtMostOneOf`] variant for all components stored in the bundle `B`.
+    #[inline]
+    pub const fn at_most_one_of() -> Self {
+        ArchetypeStatement::AtMostOneOf(PhantomData)
+    }
+
+    /// Constructs a new [`ArchetypeStatement::NoneOf`] variant for all components stored in the bundle `B`.
+    #[inline]
+    pub const fn none_of() -> Self {
+        ArchetypeStatement::NoneOf(PhantomData)
+    }
+
+    /// Constructs a new [`ArchetypeStatement::Only`] variant for all components stored in the bundle `B`.
+    #[inline]
+    pub const fn only() -> Self {
+        ArchetypeStatement::Only(PhantomData)
+    }
+}
+
+// We must pass in a generic type to all archetype statements;
+// we use the empty bundle `()` by convention.
+// These helper methods are useful because they improve type inference and consistency in user code.
+impl ArchetypeStatement<()> {
+    /// Constructs a new [`ArchetypeStatement::True`] variant.
+    #[inline]
+    pub const fn always_true() -> Self {
+        ArchetypeStatement::<()>::True
+    }
+
+    /// Constructs a new [`ArchetypeStatement::False`] variant.
+    #[inline]
+    pub const fn always_false() -> Self {
+        ArchetypeStatement::<()>::False
+    }
+}
+
+/// A type-erased version of [`ArchetypeInvariant`].
+///
+/// Intended to be used with dynamic components that cannot be represented with Rust types.
+/// Prefer [`ArchetypeInvariant`] when possible.
+#[derive(Clone, Debug, PartialEq)]
+pub struct UntypedArchetypeInvariant {
+    /// Defines which entities this invariant applies to.
+    /// This is the "if" of the if/then clause.
+    pub premise: UntypedArchetypeStatement,
+    /// Defines what must be true for the entities that this invariant applies to.
+    /// This is the "then" of the if/then clause.
+    pub consequence: UntypedArchetypeStatement,
+}
+
+impl UntypedArchetypeInvariant {
+    /// Asserts that the provided iterator of [`ComponentId`]s obeys this archetype invariant.
+    ///
+    /// `component_ids` is generally provided via the `components` field on [`Archetype`](crate::archetype::Archetype).
+    /// When testing against multiple archetypes, [`ArchetypeInvariants::test_archetype`] is preferred,
+    /// as it can more efficiently cache checks between archetypes.
+    ///
+    /// # Panics
+    /// Panics if the archetype invariant is violated.
+    pub fn test_archetype(&self, component_ids_of_archetype: impl Iterator<Item = ComponentId>) {
```

Comment:
Three reasons:
1. Since this is planned to be used for soundness, the only safe thing to do when an invariant is broken is to panic.
2. Users generally shouldn't be calling these functions directly outside of tests, since all the checking is handled automatically
3. Handling the panic logic within the testing functions lets us also handle the error formatting/reporting logic within those functions, instead of expecting calling code to implement it.

Reply:
Then I really think it should be `pub(crate)` instead if we can't imagine a public use-case for it.

Reply:
Fixed in 66910814304cf2209fdcbf6acd35491d4f288dc8 :)